<!doctype html>
<html lang="zh-CN">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width">
  <title>assert.doesNotReject(asyncFn[, error][, message]) | Node.js API 文档</title>
  <link rel="icon" sizes="32x32" type="image/png" href="http://static.nodejs.cn/_static/img/favicon.png"><link rel="stylesheet" href="/api/api.css">
  <script>
    document.write('<link href="http://nodejs.cn/api/assert/'&#32;+&#32;(location.port?'':'//static.nodejs.cn')&#32;+&#32;'/_static/css/api.css?1576756292555" rel="stylesheet">')
  </script>
</head>
<body class="alt apidoc" id="page_api_item">
  <div id="api-section-assert">
    <div id="content" class="clearfix">
  
      <div id="column1" data-id="__ID__" class="interior">
        <header>
          <h1>assert.doesNotReject(asyncFn[, error][, message])</h1>
          <div id="gtoc">
            <ul>
              <li>v12.14.0</li>
              <li>
                <a href="../assert.html#assert_assert_doesnotreject_asyncfn_error_message" name="toc">返回上层文档</a>
              </li>
              <li>
                <a href="https://github.com/nodejscn/node-api-cn/edit/master/assert/assert_doesnotreject_asyncfn_error_message.md" rel="nofollow" target="_blank">提交修改</a>
              </li>
              <li>
                <a href="http://nodejs.cn/search" name="toc">搜索</a>
              </li>
            </ul>
          </div>
          <hr>
        </header>

        <div id="biz_item"></div>
  
        <div id="apicontent">
          <div id="content_left"><div class="api_metadata">
<span>&#x65B0;&#x589E;&#x4E8E;: v10.0.0</span>
</div><ul>
<li><code>asyncFn</code> <a href="http://nodejs.cn/s/ceTQa6" target="_blank" rel="nofollow" class="type">&lt;Function&gt;</a> | <a href="http://nodejs.cn/s/ri1kj8" target="_blank" rel="nofollow" class="type">&lt;Promise&gt;</a></li>
<li><code>error</code> <a href="http://nodejs.cn/s/G38byW" target="_blank" rel="nofollow" class="type">&lt;RegExp&gt;</a> | <a href="http://nodejs.cn/s/ceTQa6" target="_blank" rel="nofollow" class="type">&lt;Function&gt;</a></li>
<li><code>message</code> <a href="http://nodejs.cn/s/9Tw2bK" target="_blank" rel="nofollow" class="type">&lt;string&gt;</a></li>
</ul><p>&#x7B49;&#x5F85; <code>asyncFn</code> Promise&#xFF0C;&#x6216;&#x8005;&#xFF0C;&#x5982;&#x679C; <code>asyncFn</code> &#x662F;&#x4E00;&#x4E2A;&#x51FD;&#x6570;&#xFF0C;&#x5219;&#x7ACB;&#x5373;&#x8C03;&#x7528;&#x8BE5;&#x51FD;&#x6570;&#x5E76;&#x7B49;&#x5F85;&#x8FD4;&#x56DE;&#x7684; Promise &#x5B8C;&#x6210;&#x3002;
&#x7136;&#x540E;&#x5B83;&#x5C06;&#x68C0;&#x67E5; Promise &#x662F;&#x5426;&#x88AB;&#x62D2;&#x7EDD;&#x3002;</p><p>&#x5982;&#x679C; <code>asyncFn</code> &#x662F;&#x4E00;&#x4E2A;&#x51FD;&#x6570;&#x5E76;&#x4E14;&#x5B83;&#x540C;&#x6B65;&#x629B;&#x51FA;&#x4E00;&#x4E2A;&#x9519;&#x8BEF;&#xFF0C;&#x5219; <code>assert.doesNotReject()</code> &#x5C06;&#x8FD4;&#x56DE;&#x4E00;&#x4E2A;&#x5E26;&#x6709;&#x8BE5;&#x9519;&#x8BEF;&#x7684;&#x88AB;&#x62D2;&#x7EDD;&#x7684; <code>Promise</code>&#x3002;
&#x5982;&#x679C;&#x51FD;&#x6570;&#x672A;&#x8FD4;&#x56DE; Promise&#xFF0C;&#x5219; <code>assert.doesNotReject()</code> &#x5C06;&#x8FD4;&#x56DE;&#x4E00;&#x4E2A;&#x88AB;&#x62D2;&#x7EDD;&#x7684; <code>Promise</code>&#xFF0C;&#x5176;&#x4E2D;&#x5305;&#x542B; <a href="http://nodejs.cn/s/zKCD5R" rel="nofollow"><code>ERR_INVALID_RETURN_VALUE</code></a> &#x9519;&#x8BEF;&#x3002;
&#x5728;&#x8FD9;&#x4E24;&#x79CD;&#x60C5;&#x51B5;&#x4E0B;&#x90FD;&#x4F1A;&#x8DF3;&#x8FC7;&#x9519;&#x8BEF;&#x5904;&#x7406;&#x51FD;&#x6570;&#x3002;</p><p>&#x4F7F;&#x7528; <code>assert.doesNotReject()</code> &#x5B9E;&#x9645;&#x4E0A;&#x6CA1;&#x6709;&#x7528;&#x5904;&#xFF0C;&#x56E0;&#x4E3A;&#x6355;&#x83B7;&#x62D2;&#x7EDD;&#x7136;&#x540E;&#x518D;&#x6B21;&#x62D2;&#x7EDD;&#x5B83;&#x51E0;&#x4E4E;&#x6CA1;&#x6709;&#x4EC0;&#x4E48;&#x597D;&#x5904;&#x3002;
&#x5E94;&#x8BE5;&#x8003;&#x8651;&#x5728;&#x4E0D;&#x5E94;&#x62D2;&#x7EDD;&#x7684;&#x7279;&#x5B9A;&#x4EE3;&#x7801;&#x8DEF;&#x5F84;&#x65C1;&#x8FB9;&#x6DFB;&#x52A0;&#x6CE8;&#x91CA;&#xFF0C;&#x5E76;&#x5C3D;&#x53EF;&#x80FD;&#x4FDD;&#x7559;&#x9519;&#x8BEF;&#x6D88;&#x606F;&#x3002;</p><p>&#x5982;&#x679C;&#x6307;&#x5B9A;&#xFF0C;&#x5219; <code>error</code> &#x53EF;&#x4EE5;&#x662F; <a href="http://nodejs.cn/s/5wDLcp" rel="nofollow"><code>Class</code></a>&#x3001;<a href="http://nodejs.cn/s/tbQJCP" rel="nofollow"><code>RegExp</code></a> &#x6216;&#x9A8C;&#x8BC1;&#x51FD;&#x6570;&#x3002;
&#x6709;&#x5173;&#x66F4;&#x591A;&#x8BE6;&#x7EC6;&#x4FE1;&#x606F;&#xFF0C;&#x8BF7;&#x53C2;&#x89C1; <a href="http://nodejs.cn/s/sGRMSp" rel="nofollow"><code>assert.throws()</code></a>&#x3002;</p><p>&#x9664;&#x4E86;&#x7B49;&#x5F85;&#x7684;&#x5F02;&#x6B65;&#x6027;&#x8D28;&#x4E4B;&#x5916;&#xFF0C;&#x5B8C;&#x6210;&#x884C;&#x4E3A;&#x4E0E; <a href="http://nodejs.cn/s/mm4aoA" rel="nofollow"><code>assert.doesNotThrow()</code></a> &#x5B8C;&#x5168;&#x76F8;&#x540C;&#x3002;</p><pre><code class="language-js">(async () =&gt; {
  await assert.doesNotReject(
    async () =&gt; {
      throw new TypeError(&apos;&#x9519;&#x8BEF;&#x503C;&apos;);
    },
    SyntaxError
  );
})();
</code></pre><pre><code class="language-js">assert.doesNotReject(Promise.reject(new TypeError(&apos;&#x9519;&#x8BEF;&#x503C;&apos;)))
  .then(() =&gt; {
    // ...
  });
</code></pre></div>
          <div id="content_right"><div class="api_metadata">
<span>Added in: v10.0.0</span>
</div><ul>
<li><code>asyncFn</code> <a href="http://nodejs.cn/s/ceTQa6" target="_blank" rel="nofollow" class="type">&lt;Function&gt;</a> | <a href="http://nodejs.cn/s/ri1kj8" target="_blank" rel="nofollow" class="type">&lt;Promise&gt;</a></li>
<li><code>error</code> <a href="http://nodejs.cn/s/G38byW" target="_blank" rel="nofollow" class="type">&lt;RegExp&gt;</a> | <a href="http://nodejs.cn/s/ceTQa6" target="_blank" rel="nofollow" class="type">&lt;Function&gt;</a></li>
<li><code>message</code> <a href="http://nodejs.cn/s/9Tw2bK" target="_blank" rel="nofollow" class="type">&lt;string&gt;</a></li>
</ul><p>Awaits the <code>asyncFn</code> promise or, if <code>asyncFn</code> is a function, immediately
calls the function and awaits the returned promise to complete. It will then
check that the promise is not rejected.</p><p>If <code>asyncFn</code> is a function and it throws an error synchronously,
<code>assert.doesNotReject()</code> will return a rejected <code>Promise</code> with that error. If
the function does not return a promise, <code>assert.doesNotReject()</code> will return a
rejected <code>Promise</code> with an <a href="http://nodejs.cn/s/zKCD5R" rel="nofollow"><code>ERR_INVALID_RETURN_VALUE</code></a> error. In both cases
the error handler is skipped.</p><p>Using <code>assert.doesNotReject()</code> is actually not useful because there is little
benefit in catching a rejection and then rejecting it again. Instead, consider
adding a comment next to the specific code path that should not reject and keep
error messages as expressive as possible.</p><p>If specified, <code>error</code> can be a <a href="http://nodejs.cn/s/5wDLcp" rel="nofollow"><code>Class</code></a>, <a href="http://nodejs.cn/s/tbQJCP" rel="nofollow"><code>RegExp</code></a> or a validation
function. See <a href="http://nodejs.cn/s/sGRMSp" rel="nofollow"><code>assert.throws()</code></a> for more details.</p><p>Besides the async nature to await the completion behaves identically to
<a href="http://nodejs.cn/s/mm4aoA" rel="nofollow"><code>assert.doesNotThrow()</code></a>.</p><pre><code class="language-js">(async () =&gt; {
  await assert.doesNotReject(
    async () =&gt; {
      throw new TypeError(&apos;Wrong value&apos;);
    },
    SyntaxError
  );
})();
</code></pre><pre><code class="language-js">assert.doesNotReject(Promise.reject(new TypeError(&apos;Wrong value&apos;)))
  .then(() =&gt; {
    // ...
  });
</code></pre></div>
          <div></div>
        </div>
      </div>
    </div>
  </div>
  <script>
    document.write('<script src="http://nodejs.cn/api/assert/'&#32;+&#32;(location.port&#32;?&#32;''&#32;:&#32;'//static.nodejs.cn')&#32;+&#32;'/_static/js/api.js?1576756292555"><\/script>');
  </script>
</body>
</html>